File: 21561 - Suvam Tamang/Week 2/1541. Minimum Insertions to Balance a Parentheses String.java
class Solution {
    public int minInsertions(String s) {
        Stack <Character> stack = new Stack<>();
        int result = 0;
        char c = ' ';
        for(int i=0;i<s.length();i++){
            c = s.charAt(i);
            // we push all the opening paranthesis to the stack
            if(c=='('){
                stack.push(c);
            }
            else if(c==')'){
                // The condition below signifies that we have found "))"
                if(i+1 < s.length() && s.charAt(i+1)==')'){
                    if(stack.empty()){
                        result += 1;
                        i += 1;
                    }
                    else{
                        stack.pop();
                        i += 1;
                    }
                }
                // else we found only a single ')'
                else{
                    if(stack.empty()){
                        result += 2;
                    }
                    else{
                        stack.pop();
                        result += 1;
                    }
                }
            }
       }
       return stack.size() * 2 + result;
    }
}

File: 21558 - Nithin/week 2/1541-minimum-insertions-to-balance-a-parentheses-string/1541-minimum-insertions-to-balance-a-parentheses-string.cpp
class Solution {
public:
    int minInsertions(string s) {
        stack<char> stck1;
        int min_ins = 0;
        int len = s.length();
        for(int i = 0; i < len; i++){
            if(s[i] == '(')
                stck1.push(s[i]);
            else{
                if(s[i+1] == ')'){
                    if(!stck1.empty())
                        stck1.pop();
                    else
                        min_ins += 1;
                    i += 1;
                }
                else
                    if(!stck1.empty()){
                        stck1.pop();
                        min_ins += 1;
                    }
                    else{
                       min_ins += 2;
                     }
                }
        }
        
        while(!stck1.empty()){
            min_ins += 2;
            stck1.pop();
        }
        return min_ins;
    }
};